### PR TITLE
feat: standardize user agent value

### PR DESCRIPTION
**Pull request by AllanZhengYP**
The purpose of this change is to standardize the user agent behavior across the AWS tools. A brief summary is:
1. Only basic version information will be sent with `User-Agent` header: `aws-sdk-js/[version]`;
1. More tracking metrics are sent with `x-amz-user-agent` header
1. User specified user agent prefix are prepended onto both headers.
1. Include *required* tool id and version: `aws-sdk-js/[version]`,
1. Include *required* language identifier and version: `lang/js`,
1. Include *optional* language metadata: `md/nodejs/[version]` or `md/browser/[version]` or `md/rn/[version]` 
1. Include *required* OS info: `os/[name]/[version]`
1. include *optional* modular SDK module version: `api/[serviceId]/[version]`
1. include *optional* client config like: `cfg/retry-mode/[mode]`

Here's an example UA string:
`aws-sdk-js/3.0.0-rc.9 os/linux/4.14.200-155.322.amzn2.x86_64 lang/js md/nodejs/12.14.0 api/iam/3.0.0-rc.9 cfg/retry-mode/standard`

In browser, I introduced [bowser](https://www.npmjs.com/package/bowser) to make best effort infer the OS and browser version from the UA string. This is so far the best choice considering downloads, size, and GH activities.

TODO: bellows are features to be addressed in following PR:
1. Paginator, waiter and lib-storage has limitation of adding `ft/[paginator, waiter, s3-transfer]` ua to these requests. We need to address the issues with these features before we can add user agent for them.

A noteworthy change in this PR is that I added a [`UserAgent` abstraction](https://github.com/AllanZhengYP/aws-sdk-js-v3/blob/a1cdcd263c1c7f33c7d1283f71e5abad05419f8e/packages/types/src/util.ts#L105-L115). So users don't need to deal with the string concatenation, instead they can just put the tuple by `context.userAgent.push(["foo-api", "1.0.0"])`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "java.project.referencedLibraries": [
+    "lib/**/*.jar",
+    "/home/ec2-user/workspace/smithy-typescript/**/*.jar",
+    "/home/ec2-user/workspace/smithy/**/*.jar"
+  ],
+  "java.checkstyle.configuration": "${workspaceFolder}/config/checkstyle/checkstyle.xml"
+}
```
Comment (AllanZhengYP) [inline codegen/.vscode/settings.json:8]: To remove this file

File: packages/middleware-user-agent/src/inject-ua-plugin.spec.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { constructStack } from "@aws-sdk/middleware-stack";
```
Comment (AllanZhengYP) [inline packages/middleware-user-agent/src/inject-ua-plugin.spec.ts:1]: This is not required for now. I can remove it and introduce back when needed.

File: packages/util-user-agent-browser/src/index.ts [hunk 3]
```diff
@@ -1,5 +1,36 @@
...
+/**
+ * Default provider to the user agent in browsers. It's a best effort to infer
+ * the device information. It uses bowser library to detect the browser and virsion
+ */
+export const defaultUserAgent = ({
+  serviceId,
+  clientVersion,
+}: DefaultUserAgentOptions): Provider<UserAgent> => async () => {
+  // TODO: remove this post GA and version changed to 3.x.x
+  clientVersion = clientVersion.replace(/^1\./, "3.");
```
Comment (AllanZhengYP) [inline packages/util-user-agent-browser/src/index.ts:15]: To remove this.

File: packages/util-user-agent-browser/src/index.native.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Default provider to the user agent in ReactNative. It's a best effort to infer
+ * the device information. It uses bowser library to detect the browser and virsion
+ */
+export const defaultUserAgent = ({
+  serviceId,
+  clientVersion,
+}: DefaultUserAgentOptions): Provider<UserAgent> => async () => {
+  // TODO: remove this post GA and version changed to 3.x.x
+  clientVersion = clientVersion.replace(/^1\./, "3.");
```
Comment (AllanZhengYP) [inline packages/util-user-agent-browser/src/index.native.ts:14]: to remove this.

File: packages/util-user-agent-node/src/index.ts [hunk 5]
```diff
@@ -1,9 +1,55 @@
...
+
+/**
+ * Collect metrics from runtime to put into user agent.
+ */
+export const defaultUserAgent = ({
+  serviceId,
+  clientVersion,
+}: DefaultUserAgentOptions): Provider<UserAgent> => async () => {
+  // TODO: remove this post GA and version changed to 3.x.x
+  clientVersion = clientVersion.replace(/^1\./, "3.");
```
Comment (AllanZhengYP) [inline packages/util-user-agent-node/src/index.ts:22]: to remove it.

**Review by trivikr — APPROVED** [review]

**Comment by trivikr** [comment]
Revisiting this PR to check User Agent values before and after:

Node.js:
* Before: `aws-sdk-nodejs-v3-@aws-sdk/client-dynamodb/1.0.0-rc.10 darwin/v14.17.1`
* After: `aws-sdk-js/3.1.0 os/darwin/20.5.0 lang/js md/nodejs/14.17.1 api/dynamodb/3.1.0`

Browser:
* Before: `aws-sdk-js-v3-@aws-sdk/client-dynamodb/1.0.0-rc.10 Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36`
* After: `aws-sdk-js/3.1.0 os/macOS/10.15.7 lang/js md/browser/Chrome_91.0.4472.114 api/dynamodb/3.1.0`

**Pull request merged**